src/mpid/ch4/netmod/ofi/ofi_am_impl.c

```diff
@@ -16,79 +17,104 @@ void MPIDI_OFI_deferred_am_isend_dequeue(MPIDI_OFI_deferred_am_isend_req_t * dre
+    MPL_DBG_MSG_FMT(MPIDI_CH4_DBG_GENERAL, VERBOSE,
+                    (MPL_DBG_FDEST, "try issue deferred am isend dreq=%p", dreq));
 
-    MPL_pointer_attr_t attr;
-    MPIR_GPU_query_pointer_attr(dreq->buf, &attr);
-    if (attr.type == MPL_GPU_POINTER_DEV && !MPIDI_OFI_ENABLE_HMEM) {
-        /* Force packing of GPU buffer in host memory */
-        need_packing = true;
+    if (dreq->is_pipeline) {
+        offset = MPIDIG_send_get_offset(dreq->sreq);
+        send_size = MPIDI_OFI_DEFAULT_SHORT_SEND_SIZE - (offset ? 0 : dreq->am_hdr_sz)
+            - sizeof(MPIDI_OFI_am_header_t);
+        send_size = MPL_MIN(send_size, MPIDIG_send_get_data_sz_left(dreq->sreq));
+    } else {
+        send_size = dreq->data_sz;
+        offset = 0;
     }
 
-    if (need_packing) {
-        /* FIXME: currently we always do packing, also for high density types. However,
-         * we should not do packing unless needed. Also, for large low-density types
-         * we should not allocate the entire buffer and do the packing at once. */
-        /* TODO: (1) Skip packing for high-density datatypes;
-         *       (2) Pipeline allocation for low-density datatypes; */
-        if (do_eager) {
-            mpi_errno = MPIDU_genq_private_pool_alloc_cell(MPIDI_OFI_global.pack_buf_pool,
-                                                           (void **) &send_buf);
-            if (send_buf == NULL) {
-                /* no buffer available, suppress error and exit */
-                mpi_errno = MPI_SUCCESS;
-                goto fn_exit;
-            }
-            MPIR_ERR_CHECK(mpi_errno);
-        } else {
-            /* only for large message RDMA read */
-            MPL_gpu_malloc_host((void **) &send_buf, data_sz);
+    if (dreq->need_packing) {
+        mpi_errno = MPIDU_genq_private_pool_alloc_cell(MPIDI_OFI_global.pack_buf_pool,
+                                                       (void **) &send_buf);
+        if (send_buf == NULL) {
+            /* no buffer available, suppress error and exit */
+            mpi_errno = MPI_SUCCESS;
+            goto fn_exit;
         }
-        mpi_errno = MPIR_Typerep_pack(dreq->buf, dreq->count, dreq->datatype, 0, send_buf,
-                                      dreq->data_sz, &last);
         MPIR_ERR_CHECK(mpi_errno);
+        mpi_errno = MPIR_Typerep_pack(dreq->buf, dreq->count, dreq->datatype, offset, send_buf,
+                                      send_size, &packed_size);
+        MPIR_ERR_CHECK(mpi_errno);
+        send_size = packed_size;
 
-        MPIDI_OFI_AMREQUEST_HDR(dreq->sreq, pack_buffer) = send_buf;
+        MPIDI_OFI_am_create_send_request(dreq->sreq, dreq->am_hdr, dreq->am_hdr_sz, &send_req);
+        send_req->req_hdr.pack_buffer = send_buf;
     } else {
-        MPIDI_OFI_AMREQUEST_HDR(dreq->sreq, pack_buffer) = NULL;
+        MPIDI_Datatype_get_ptr_lb(dreq->count, dreq->datatype, dt_ptr, dt_true_lb);
```

Comment:
Add comment that we are assuming contiguous datatype here and leave `TODO` as to deal with unpacked non-contig data.

Reply:
ok